### ISSUE TITLE
Override snippet or explicitly disable snippets file.

### ISSUE THREAD

**Issue by jbactad** [neoclide/coc-snippets#141]
Hi guys,

I just want to ask if there is a way to override a snippet suggestion so that suggestions don't appear multiple times in the suggestion popup.

If this is not possible, maybe there is a way to ignore a snippets file from an extension or directory?

Thanks in advance!

**Comment by chemzqm** [comment]
No, no such kind of feature

**Comment by wfxr** [comment]
`ultisnippets` support overrides the snippets by setting priority in snippet file like this:

```snippet
# go.snippets
priority 100

snippet go "go func() { ... }()" b,m
go func() {
	${1:${VISUAL}}
}()
endsnippet
```

It should override the snippet `go` provided by `vim-snippets` which priority is 0. But seems `coc-snippets` ignored the priority and always show all of them (even the `coc-snippets`'s `go` snippet always ranks first).

![image](https://user-images.githubusercontent.com/6105425/85646869-b470d380-b6cf-11ea-9495-ac3bede801c5.png)
(The last one is the snippet defined above with priority 100)

Since no override feature, is there anyway to make custom snippets rank higher than the defaults? @chemzqm

**Comment by nomasprime** [comment]
When using UltiSnips it overrides other snippets, see [here](https://github.com/SirVer/ultisnips/issues/364#issuecomment-53284215).

This can't be done in coc-snippets?

**Comment by chemzqm** [comment]
It's possible to use priority as well.

**Comment by nomasprime** [comment]
@chemzqm does this work with snipMate snippets though in coc-snippets? I tried using a lower priority for the snipMate snippet and I'm still asked to choose.

**Comment by chemzqm** [comment]
Priority only work with Ultisnip snippet for now, and you still need to make choice.

**Comment by nomasprime** [comment]
Thanks for clarifying @chemzqm.

I'm going to stick with my own fork of vim-snippets for now anyway because I prefer to use my own naming conventions.

**Comment by unphased** [comment]
I have a use case for being able to disable certain snippets. Typically I'd agree with the more the merrier but there's some stuff you just know you're not going to want and don't want them to clutter up the list. 

Prime example right now is `javascript-jquery.snippets`. I do not and will not use jquery so please let me nuke them.

**Comment by chemzqm** [comment]
Use configuration "snippets.excludePatterns"

**Comment by unphased** [comment]
@chemzqm The doc states, 

> `snippets.excludePatterns`: List of filepath patterns to exclude, $HOME and other environment variables are expanded.

1. It's not clear either from this description, nor from the coc-settings.json built-in helper popup 
    >  List of minimatch patterns for filepath to exclude, support expand homedir and environment variables.
    what this option really means. Is it a list of patterns for which if the file we're editing matches that the snippets will all be disabled? Is this supposed to be a list of lists or list of objects where we can specify for a given filepath pattern a list of snippet types to enable on it? (that might be useful)
2. Regardless of how it works what we want is to control whether specific snippet definitions are loaded or not. I want to never ever use `javascript-jquery.snippets`, and `snippets.excludePatterns` does not assist with this need.
3. I have tested the following in `coc-settings.json` and it did nothing: `"snippets.excludePatterns": ["*jquery*"]`
4. I tried setting the config to this: `"snippets.excludePatterns": ["*nosnippets*"]` and then editing a file `nvim nosnippets.file`, and it did not prevent snippets from popping up either.

**Comment by unphased** [comment]
All that being said --- I am finding `:CocCommand workspace.showOutput` for snippets is producing helpful outputt which shows that it is my use of `vim-snippets` plugin that is responsible for e.g. the jquery stuff i'm seeing which means I can nuke it from there to take care of the situation as well as use this to further troubleshoot which specific snippet files i want to load and see if I can't make sense out of how the `excludePatterns` config is supposed to work.

**Comment by chemzqm** [comment]
The snippet file won't be loaded when any of the pattern in the list matched.

**Comment by unphased** [comment]
@chemzqm I have gotten pretty far instrumenting the `~/.config/coc/extensions/node_modules/coc-snippets/lib/index.js` file. However this is a ts-generated js file and not tremendously suitable for debugging work. Could you tell me how I can point coc to load a cloned coc-snippets repo in which I can edit? Should I use `npm link` from `~/.config/coc/extensions/`?

So far this is a line I added that gives me some interesting results. Look: 

```
isIgnored(filepath) {
  let ignored = false;
  let { excludes } = this.config;
  for (let p of excludes) {
    fs.appendFileSync('/tmp/coc_snippets_debug.log', `- minimatch test debug filepath=${filepath} p=${p} import_minimatch.default=>>>${import_minimatch.default}<<<
    → RESULT=${(0, import_minimatch.default)(filepath, p, { dot: true })}\n`);
    if ((0, import_minimatch.default)(filepath, p, { dot: true })) {
      ignored = true;
      this.info(`File ignored by excludePatterns: ${filepath}`);
      break;
    }
  }
  return ignored;
}
```
That `appendFileSync` line is my poor mans debug statement. Hey it works. 

The output is: 
```
- minimatch test debug filepath=/home/slu/.vim/plugged/vim-snippets/snippets/javascript/javascript-jquery.snippets p=*jquery.snippets import_minimatch.default=>>>(p, pattern, options = {}) => {
      assertValidPattern(pattern);
      if (!options.nocomment && pattern.charAt(0) === "#") {
        return false;
      }
      return new Minimatch(pattern, options).match(p);
    }<<< RESULT=false
```

This shows me a number of things

- the minimatch seems to fail matching `/home/slu/.vim/plugged/vim-snippets/snippets/javascript/javascript-jquery.snippets` via the pattern `*jquery.snippets`. I cannot fathom why. 
- The anonymous function body printed here is merely the exported entrypoint of minimatch.js dependency. Nothing looks weird there. 

I must be missing something really basic.

**Comment by unphased** [comment]
Great! OK, I got to the bottom of it. The last post was a good rubber duck session. 

Reading the minimatch doc some more I realized I must use a pattern like `**/*jquery.snippets`. Indeed this did the trick. 

    [Info  - 07:39:14.206] File ignored by excludePatterns: /home/slu/.vim/plugged/vim-snippets/snippets/javascript/javascript-jquery.snippets
    [Info  - 07:39:14.206] File ignored by excludePatterns: /home/slu/.vim/plugged/vim-snippets/snippets/javascript/javascript-jquery.snippets

Yes this finally gives me what I want and @chemzqm you are vindicated as it works exactly as you say. 

I have one final question though. Why does the message print twice? I noticed also that my debug statement from my debug session above also printed twice. Doesn't seem to hurt much. But it should not print twice.
